lib/Dialect/AIG/Analysis/LongestPathAnalysis.cpp

```diff
@@ -172,47 +262,1100 @@ struct llvm::DenseMapInfo<Object> {
+                      SmallVectorImpl<OpenPath> &results) {
+    return markFanIn(op, bitPos, results);
+  }
+
+  LogicalResult visitDefault(Operation *op, size_t bitPos,
+                             SmallVectorImpl<OpenPath> &results);
+
+  // Helper functions.
+  LogicalResult addEdge(Value to, size_t toBitPos, int64_t delay,
+                        SmallVectorImpl<OpenPath> &results);
+  LogicalResult markFanIn(Value value, size_t bitPos,
+                          SmallVectorImpl<OpenPath> &results);
+  LogicalResult markRegFanOut(Value fanOut, Value start, Value reset = {},
+                              Value resetValue = {}, Value enable = {});
+
+  // The module we are visiting.
+  hw::HWModuleOp module;
+  Context *ctx;
+
+  // Thread-local data structures.
+  std::unique_ptr<circt::igraph::InstancePathCache> instancePathCache;
+  // TODO: Make debug points optional.
+  std::unique_ptr<llvm::ImmutableListFactory<DebugPoint>> debugPointFactory;
+
+  // A map from the value point to the longest paths.
+  DenseMap<std::pair<Value, size_t>, SmallVector<OpenPath>> cachedResults;
+
+  // A map from the object to the longest paths.
+  DenseMap<Object, SmallVector<OpenPath>> fanOutResults;
+
+  // This is set to true when the thread is done.
+  std::atomic_bool done;
+
+  // A flag to indicate the module is top-level.
+  bool topLevel = false;
+};
+
+LocalVisitor::LocalVisitor(hw::HWModuleOp module, Context *ctx)
+    : module(module), ctx(ctx) {
+  debugPointFactory =
+      std::make_unique<llvm::ImmutableListFactory<DebugPoint>>();
+  instancePathCache = ctx->instanceGraph
+                          ? std::make_unique<circt::igraph::InstancePathCache>(
+                                *ctx->instanceGraph)
+                          : nullptr;
+  done = false;
+}
+
+ArrayRef<OpenPath> LocalVisitor::getResults(Value value, size_t bitPos) const {
+  std::pair<Value, size_t> valueAndBitPos(value, bitPos);
+  auto leader = ec.findLeader(valueAndBitPos);
+  if (leader != ec.member_end()) {
+    if (*leader != valueAndBitPos) {
+      // If this is not the leader, then use the leader.
+      auto root = ecMap.at(*leader);
+      return getResults(root.first, root.second);
+    }
+  }
+
+  auto it = cachedResults.find(valueAndBitPos);
+  // If not found, then consider it to be a constant.
+  if (it == cachedResults.end())
+    return {};
+  return it->second;
+}
+
+void LocalVisitor::putUnclosedResult(const Object &object, int64_t delay,
+                                     llvm::ImmutableList<DebugPoint> history,
+                                     ObjectToMaxDistance &objectToMaxDistance) {
+  auto &slot = objectToMaxDistance[object];
+  if (slot.first >= delay && delay != 0)
+    return;
+  slot = {delay, history};
+}
+
+void LocalVisitor::waitUntilDone() const {
+  // Wait for the thread to finish.
+  while (!done.load())
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
Might be worth using a condition variable here to avoid sleeping. Not sure how much of a performance implication this has in practice :slightly_smiling_face: